exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -158,38 +165,51 @@ public DrillbitContext getContext() {
     return dContext;
   }
 
-  private ExtendedLatch exitLatch = null; // used to wait to exit when things are still running
-
   /**
    * Waits until it is safe to exit. Blocks until all currently running fragments have completed.
-   *
-   * <p>This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.</p>
+   * This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.
    */
   public void waitToExit() {
-    synchronized(this) {
-      if (queries.isEmpty() && runningFragments.isEmpty()) {
-        return;
+    final long startTime = System.currentTimeMillis();
+
+    try {
+      exitLock.lock();
+      long diff;
+      while ((diff = (System.currentTimeMillis() - startTime)) < EXIT_TIMEOUT) {
+        if (queries.isEmpty() && runningFragments.isEmpty()) {
+          break;
+        }
+
+        try {
+          final boolean success = exitCondition.await(EXIT_TIMEOUT - diff, TimeUnit.MILLISECONDS);
+
+          if (!success) {
+            logger.warn("Timed out after %d millis while waiting to exit.", EXIT_TIMEOUT);
+            exitLock.lock();
```

Comment:
Why is this lock necessary?

Reply:
If the await times out, the condition variable does not automatically reacquire the lock for us, so we must do it manually.